wit/outgoing-handler.wit

```diff
@@ -0,0 +1,18 @@
+// The `wasi:http/outgoing-handler` interface is meant to be imported by
+// components and implemented by the host.
+//
+//   NOTE: in Preview3, this interface will be merged with
+//   `wasi:http/outgoing-handler` into a single `wasi:http/handler` interface
+//   that takes a `request` parameter and returns a `response` result.
+//
+default interface outgoing-handler {
+  use pkg.types.{outgoing-request, request-options, future-incoming-response}
+
+  // The parameter and result types of the `handle` function allow the caller
+  // to concurrently stream the bodies of the outgoing request and the incoming
+  // response.
+  handle: func(
+    request: outgoing-request,
+    options: option<request-options>
+  ) -> future-incoming-response
+}
```

Comment:
I find those pairings (`incoming-request` and `outgoing-response`, and `outgoing-request` and `incoming-response`) very non-intuitive and a bit redundant.

Could we consider the same prefix on the same connection, e.g. `downstream-request` and `downstream-response`, and `upstream-request` and `upstream-response`?

Reply:
RFC 9110 suggests `inbound` and `outbound` as the preferred terms here, and defines `upstream` and `downstream` differently than you are using them: https://www.rfc-editor.org/rfc/rfc9110#name-intermediaries

Reply:
Right, but the popular implementations (e.g. NGINX and Envoy) disagree with the RFC about `upstream` and `downstream`, and I'm not aware of anything that actually uses the RFC terms.

I considered suggesting `inbound` and `outbound`, but it's yet another term, and it's actually used in service mesh environment to describe traffic direction in a global sense, and not from the perspective of a local endpoint.

Reply:
Our experience at fastly was that `upstream` and `downstream` were so confusing internally, and to our customers, that we stopped using those terms throughout our api and docs.

Reply:
What did you end up replacing it with? I'm not a fan of `downstream` / `upstream` myself (but mostly because the RFCs disagree with implementations), so I'm fine with other terms.

Reply:
Agreed that upstream/downstream is extremely confusing in practice (as evidenced by my own misuse of it below).  I saw incoming/outgoing in one of the other proposals and it seemed more clear when you think of it from the component author's perspective.  That being said, note that all `incoming-`/`outgoing-` prefixes would go away in Preview3 when we get to merge all these resources/interfaces.

Reply:
Most programming languages have standardized on request/response without incoming/outgoing b/c the context is pretty obvious from whether you are implementing a server or making a client call and programmers don't really get confused about that.

I think we should delete incoming/outgoing or downstream/upstream and just stick to request/response with the context providing the direction.

Reply:
That is definitely the goal with Preview 3.  However, in a Preview 2 timeframe, where we want to do concurrent streaming of requests/responses *and* have blocking `poll_oneoff`, the concrete signatures of the request/response resources have to be different depending on whether you are reading from them or writing to them, so we unfortunately can't wait and infer when the request or response hits a call boundary.  I do believe it's possible to implement source-language libraries which are direction-agnostic (e.g., JS [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response)) *in terms of* these these directional resource types, though, so that this is abstracted from the developer (e.g., when constructing a `Response`, the impl would internally call `new-outgoing-response`...).  Avoiding this extra work and complexity is one of the main motivations for adding `future`/`stream` (and the stack-switching required to implement these in the underlying implementation).